src/scale_impls/encode.rs

```diff
@@ -80,82 +83,108 @@ impl<T> EncodeAsFields for Composite<T> {
-                let arr_ty = array.type_param.id;
-                if value.len() != array.len as usize {
-                    return Err(Error::new(ErrorKind::WrongLength {
-                        actual_len: value.len(),
-                        expected_len: array.len as usize,
-                    }));
-                }
-
-                for (idx, val) in value.values().enumerate() {
-                    val.encode_as_type_to(arr_ty, types, out).map_err(|e| e.at_idx(idx))?;
-                }
-                Ok(())
+            Ok(())
+        };
+
+        let visit_array = |out: &mut Vec<u8>,
+                           type_id: &<R as TypeResolver>::TypeId,
+                           array_len: usize|
+         -> Result<(), Error> {
+            if value.len() != array_len {
+                return Err(Error::new(ErrorKind::WrongLength {
+                    actual_len: value.len(),
+                    expected_len: array_len,
+                }));
             }
-            TypeDef::BitSequence(seq) => {
-                encode_vals_to_bitsequence(value.values(), seq, types, out)
+            for (idx, val) in value.values().enumerate() {
+                val.encode_as_type_to(type_id, types, out).map_err(|e| e.at_idx(idx))?;
             }
-            // For other types, skip our value past a 1-value composite and try again, else error.
-            _ => {
-                let mut values = value.values();
-                match (values.next(), values.next()) {
-                    // Exactly one value:
-                    (Some(value), None) => value.encode_as_type_to(type_id, types, out),
-                    // Some other number of values:
-                    _ => Err(Error::new(ErrorKind::WrongShape {
-                        actual: Kind::Tuple,
-                        expected: type_id,
-                    })),
-                }
+            Ok(())
+        };
+
+        let visit_bit_sequence =
+            |out: &mut Vec<u8>,
+             store: scale_type_resolver::BitsStoreFormat,
+             order: scale_type_resolver::BitsOrderFormat| {
+                let format = scale_bits::Format { store, order };
+                encode_vals_to_bitsequence(value.values(), out, format)
+            };
+
+        let visitor =
+            scale_type_resolver::visitor::new::<'_, &mut Vec<u8>, R::TypeId, Result<(), Error>, _>(
+                out,
+                visit_unhandled,
+            )
+            .visit_composite(|out, _| visit_composite_or_tuple(out))
+            .visit_tuple(|out, _| visit_composite_or_tuple(out))
+            .visit_sequence(visit_sequence)
+            .visit_array(visit_array)
+            .visit_bit_sequence(visit_bit_sequence);
+
+        match types.resolve_type(type_id, visitor) {
+            Ok(Ok(())) => Ok(()),
+            Ok(Err(err)) => Err(err),
+            Err(err) => {
+                Err(Error::new(ErrorKind::Custom(Box::new(StringError(format!("{:?}", err))))))
```

Comment:
How abotu using `ErrorKind::TypeResolvingError(format("{err:?}"))` instead, since that's already the variant used to capture these underlying type resolve errors; saves needing to have the custom StringError also :)